### ISSUE TITLE
Pattern inside SVG not influenced by SVG offset inside PDF, unexpected point of origin

### ISSUE THREAD

**Issue by DarSim2** [yWorks/svg2pdf.js#212]
I have a SVG containing a pattern and a rect which is filled with said pattern:

```
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:svgjs="http://svgjs.dev/svgjs" width="200" height="200">
    <defs>
        <pattern x="0" y="0" width="20" height="20" patternUnits="userSpaceOnUse" id="SvgjsPattern1003">
            <rect width="10" height="10" fill="purple" x="0" y="0"></rect>
            <rect width="10" height="10" fill="pink" x="10" y="10"></rect>
        </pattern>
    </defs>
    <rect width="75" height="75" fill="url('#SvgjsPattern1003')"></rect>
</svg>
```

When I convert  to PDF with a x and/or y offset  then the pattern doesn't inherit the offset. It doesn't start at (0, 0) of the SVG inside the PDF but apparently the pattern starts at (0, 0) of the PDF. 
```
const pdfDoc = new jsPDF({ orientation: 'l', unit: 'pt', format: 'a4' })
pdfDoc.svg(nested.node, {
                x: 17,
                y: 32,
            })
            .then(() => {
                pdfDoc.save('testStuffPDF')
            })
```

And when I print the PDF then the printer adds some margins of its own which offsets the pattern again so that the printed element filled with the pattern doesn't look the same as its counterpart on the screen.

I know that the pattern inside a SVG always starts at (0, 0) of the SVG itself. 
This is the behavior I would also expect to be seen in the PDF. Pattern should start at the origin of the SVG inside the PDF not the origin of the PDF.

[pattern_no_offset.pdf](https://github.com/yWorks/svg2pdf.js/files/8582769/pattern_no_offset.pdf)
[pattern_with_offset.pdf](https://github.com/yWorks/svg2pdf.js/files/8582771/pattern_with_offset.pdf)

**Comment by HackbrettXXX** [comment]
Thanks for the bug report. I was able to reproduce it.

**Comment by DarSim2** [comment]
Hi, I just wanted to ask, if there is any development in this matter?

Because this became kind of a pressing matter for me, so I forked svg2pdf and tried to analyze the issue with a collegue.
We didn't manage to find a fix. Our current observations are the following, maybe this will help you:

1. BoundingBox in pattern.ts is calculated correctly (for the pattern element)
2. Our guess is that the rendered children of the pattern in their render() call do not take the offset of the pattern parent element into account

**Comment by yGuy** [comment]
Unfortunately, there was no progress, here, so far. We would have removed the "help wanted" tag if we were working on the task. HackbrettXXX is not available at the moment and won't be for the next couple of months, as far as I can tell.  I am sorry to say, so we depend on the community to step in. 

Is this a problem with the way we interpret the SVG or is this a problem with the way js2pdf writes the PDF? Have you been able to narrow down the issue?

What are the corresponding involved specs for the behavior?

**Comment by DanielBretzigheimer** [comment]
I have looked into the issue and I am trying to fix it. In my research, I have found some other issues:
1. using a unit other than pt will scale the pattern elements wrong
In the following sample the red rectange has the same size as the blue rectangles in the svg (20 x 20). With unit 'pt' this works correct, but when changed to e.g. 'mm' they scale differently.
SVG:
![image](https://user-images.githubusercontent.com/23380519/184172035-46cb70e4-ded1-498a-883d-d8a639288d68.png)
PDF:
![image](https://user-images.githubusercontent.com/23380519/184172088-f9f3a285-5dc8-4daf-8c46-057238eef9bb.png)


2. adding an offset to the pattern will result in differences between svg and pdf (see [here](http://raw.githack.com/yWorks/svg2pdf.js/master/index.html?svg=%253Csvg%2520height%253D%2522200%2522%2520width%253D%2522200%2522%2520xmlns%253Axlink%253D%2522http%253A%252F%252Fwww.w3.org%252F1999%252Fxlink%2522%2520version%253D%25221.1%2522%2520xmlns%253D%2522http%253A%252F%252Fwww.w3.org%252F2000%252Fsvg%2522%2520id%253D%2522svg-to-export%2522%253E%250A%2520%2520%253Cdefs%253E%250A%2520%2520%2520%2520%253Cpattern%2520id%253D%2522pattern%2522%2520patternUnits%253D%2522userSpaceOnUse%2522%2520height%253D%252225%2522%2520width%253D%252225%2522%2520y%253D%252215%2522%2520x%253D%252215%2522%253E%250A%2520%2520%2520%2520%2520%2520%253Crect%2520y%253D%25220%2522%2520x%253D%25220%2522%2520fill%253D%2522blue%2522%2520height%253D%252220%2522%2520width%253D%252220%2522%253E%253C%252Frect%253E%250A%2520%2520%2520%2520%253C%252Fpattern%253E%250A%2520%2520%253C%252Fdefs%253E%250A%2520%2520%253Crect%2520fill%253D%2522url%28%27%2523pattern%27%29%2522%2520height%253D%2522100%2522%2520width%253D%2522100%2522%253E%253C%252Frect%253E%250A%253C%252Fsvg%253E))
![image](https://user-images.githubusercontent.com/23380519/184171333-8aa016d8-0eec-486f-b49a-ea8115bef9be.png)
PDF:
![image](https://user-images.githubusercontent.com/23380519/184171360-6000c887-a4ca-4242-bfe1-3cc9a7c05cf8.png)

I'll find these bugs and create a PR which hopefully resolves all these issues.

**Comment by DanielBretzigheimer** [comment]
After further investigation, we found a solution for the scaling problem in units besides `pt`. We had to apply the `context.pdf.internal.scaleFactor` to the `TilingPattern` and the child transform. This worked, but showed the next issue we couldn't solve.

The tiling pattern doesn't start at `y = 0` and looks to be offset a little. We speculated that this could be from text baseline or something like that. If you have any idea how this could be fixed, please tell us. We will look further into the issue.

SVG:
![image](https://user-images.githubusercontent.com/23380519/184452217-589ae9b5-cedd-4eff-b179-973b7dce4bce.png)

PDF:
![image](https://user-images.githubusercontent.com/23380519/184452201-df95d230-5241-4a2f-bafd-961d30a99ed6.png)

**Comment by DarSim2** [comment]
I tried the fix from @DanielBretzigheimer 's PR, it is working 😄

**Comment by DarSim2** [comment]
I found another related problem with the pattern in PDF, it also doesn't consider viewbox values of the svg. It just alsways shows the same part of the pattern as if there was no viewbox on the svg

**Comment by yGuy** [comment]
That's not true. For one, there have been improvements (check the commits above) and secondly "not possible to use any type of pattern" is wrong. Many patterns do work. Which one specifically does not work for you? Can you share a repro?

**Comment by Shiuyin** [comment]
Thanks for your answer.

1. I know the commits above, i helped DanielBretzigheimer working on them, he is a colleague of mine.
2. These commits have not been merged into the official lib.
3. These commits still do not adress open issues with patterns as DarSim2 has pointed out (viewbox handling) and we were unable to fix them on our own when we tried.
4. Example code is in the original issue, just add a viewbox to it.

**Comment by yGuy** [comment]
OK. So to summarize: there are still cases that don't work for you, but the feature does work for many other patterns. You helped try to solve the issue (thanks!) but you were unsuccessful and need help. That's what the tag at this issue says. So I guess the answer to your question is: no, there aren't any updates, sorry. If you can find someone who can fix it for you, we will be happy to accept a PR. If this is super important to you, you could open a bounty and have someone work on it.

**Comment by Shiuyin** [comment]
Correct. I hope to be able to look into it myself in a few weeks again.
